Generate annotations for NewType types (fixes #22)

folz opened the pull request:
Per #22, monkeytype generates invalid type signatures for NewType types because they are valid mypy annotations but `new_type()` gets inspected as a function at runtime. This PR special-cases support for objects with `__supertype__` to return their `__name__` instead of their repr, based on https://github.com/python/cpython/blob/v3.6.3/Lib/typing.py#L2227

Since a comment says `render_annotation` was derived from `inspect.formatannotation`, I'll point out that this PR causes render_annotation to disagree with inspect.formatannotation. I think this is okay in theory since render_annotation is used for mypy types while formatannotation is used for runtime inspection, but I'd appreciate a second opinion. For what it's worth, this patch allowed me to use monkeytype against my NewType-using codebase successfully.

This seems important enough to warrant tests - if you think this PR is taking a good approach, I can write tests for it as well.

carljm commented:
This approach looks fine to me; thanks for the report and fix!

Normally we try to automatically provide the right imports for annotations too, so `monkeytype apply` doesn't result in modules with missing imports. However, I don't think that's possible with the current runtime implementation of `NewType`; the calling module is not stored anywhere in the resulting function object, so there's no way we can introspect it. So unless we change that in `typing.py` we're stuck with the limitation of not being able to automatically add NewType imports.

It would be great to add tests for NewType in `test_stubs.py`; thanks!

carljm commented:
Er, never mind about generating imports; `NewType` annotations can only occur as existing annotations, never as traced types. So the import should always be present already.

folz commented:
Added a test for this and updated the changelog. I think you're right that this doesn't need to worry about imports - if the import isn't present, it wasn't valid to mypy in the first place.

carljm commented:
Awesome, thank you!
